addons/l10n_in_edi/models/account_edi_format.py

```diff
@@ -0,0 +1,556 @@
+            if self._l10n_in_edi_is_test_mode():
+                _test_logger.info("skip invoice check configuration for Indian EDI and send to server")
+                return False
+            return invoice.is_sale_document() and invoice.country_code == 'IN' and invoice.l10n_in_gst_treatment in (
+                "regular",
+                "composition",
+                "overseas",
+                "special_economic_zone",
+                "deemed_export",
+            )
+        return super()._is_required_for_invoice(invoice)
+
+    def _needs_web_services(self):
+        self.ensure_one()
+        return self.code == "in_einvoice_1_03" or super()._needs_web_services()
+
+    def _get_invoice_edi_content(self, move):
+        if self.code != "in_einvoice_1_03":
+            return super()._get_invoice_edi_content(move)
+        json_dump = json.dumps(self._l10n_in_edi_generate_json(move))
+        return json_dump.encode()
+
+    def _l10n_in_edi_extract_digits(self, string):
+        if not string:
+            return string
+        matches = re.findall(r"\d+", string)
+        result = "".join(matches)
+        return result
+
+    def _check_move_configuration(self, move):
+        if self.code != "in_einvoice_1_03":
+            return super()._check_move_configuration(move)
+        error_message = []
+        error_message += self._l10n_in_validate_partner(move.partner_id)
+        error_message += self._l10n_in_validate_partner(move.company_id.partner_id, is_company=True)
+        if not move.company_id.sudo().l10n_in_edi_username or not move.company_id.sudo().l10n_in_edi_password:
+            error_message.append(_(
+                "A username and password still needs to be set for the E-invoice(IN).\n"
+                "It needs to be added in the Settings."))
+        if not re.match("^.{1,16}$", move.name):
+            error_message.append(_("Invoice number should not be more than 16 characters"))
+        for line in move.invoice_line_ids.filtered(lambda line: not line.display_type):
+            if line.product_id:
+                if not line.product_id.l10n_in_hsn_code:
+                    error_message.append(_("HSN code is not set in product %s", line.product_id.name))
+                elif not re.match("^[0-9]+$", line.product_id.l10n_in_hsn_code):
+                    error_message.append(_(
+                        "Invalid HSN Code (%s) in product %s", line.product_id.l10n_in_hsn_code, line.product_id.name
+                    ))
+            else:
+                error_message.append(_("product is required to get HSN code"))
+        return error_message
+
+    def _l10n_in_edi_get_iap_buy_credits_message(self, company):
+        base_url = "https://iap-sandbox.odoo.com/iap/1/credit" if not company.sudo().l10n_in_edi_production_env else ""
+        url = self.env["iap.account"].get_credits_url(service_name="l10n_in_edi", base_url=base_url)
+        return """<p><b>%s</b></p><p>%s</p>""" % (
```

Comment:
Can you use markup?

Reply:
yes, let me check

Reply:
I use Markup it gives security risk!

Reply:
@jva-odoo security risk does not mean "forbidden", it means "danger zone", we ask a review to make sure it's correct but it's actually better than crafting html by hand (which is difficult to detect via regex but is definitly in the "danger zone" too, just undetected.

Reply:
okay.